Dockerfile

```diff
@@ -12,52 +12,57 @@
 # runtime-image
 #   - Copies the virtual environment into place.
 #   - Runs a non-root user.
-#   - Configures gunicorn.
+#   - Sets up the entrypoint and port.
 
-FROM tiangolo/uvicorn-gunicorn:python3.8-slim AS base-image
+FROM python:3.10.2-slim-bullseye as base-image
 
 # Update system packages
 COPY scripts/install-base-packages.sh .
-RUN ./install-base-packages.sh
-RUN rm ./install-base-packages.sh
+RUN ./install-base-packages.sh && rm ./install-base-packages.sh
 
 FROM base-image AS dependencies-image
 
+# Install system packages only needed for building dependencies.
+COPY scripts/install-dependency-packages.sh .
+RUN ./install-dependency-packages.sh
+
+# Create a Python virtual environment
+ENV VIRTUAL_ENV=/opt/venv
+RUN python -m venv $VIRTUAL_ENV
+# Make sure we use the virtualenv
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 # Put the latest pip and setuptools in the virtualenv
 RUN pip install --upgrade --no-cache-dir pip setuptools wheel
 
 # Install the app's Python runtime dependencies
 COPY requirements/main.txt ./requirements.txt
 RUN pip install --quiet --no-cache-dir -r requirements.txt
 
-FROM dependencies-image AS runtime-image
+FROM dependencies-image AS install-image
+
+# Use the virtualenv
+ENV PATH="/opt/venv/bin:$PATH"
 
-# Install the web application
-COPY . /app
-RUN pip install --no-cache-dir /app
+COPY . /workdir
+WORKDIR /workdir
+RUN pip install --no-cache-dir .
+
+FROM base-image AS runtime-image
 
 # Create a non-root user
 RUN useradd --create-home appuser
-WORKDIR /home/appuser
 
-# Switch to non-root user
-USER appuser
+# Copy the virtualenv
+COPY --from=install-image /opt/venv /opt/venv
 
-# Copy the test butler registries to allow us to run the application with it
-COPY tests/data/LSSTCam LSSTCam
-COPY tests/data/LATISS LATISS
+# Make sure we use the virtualenv
+ENV PATH="/opt/venv/bin:$PATH"
 
-# We use a module name other than app, so tell the base image that.  This
-# does not copy the app into /app as is recommended by the base Docker
-# image documentation and instead relies on the module search path as
-# modified by the virtualenv.
-ENV MODULE_NAME=exposurelog.main
-
-# The default starts 40 workers, which exhausts the available connections
-# on a micro Cloud SQL PostgreSQL server and seems excessive since we can
-# scale with Kubernetes.  Cap the workers at 10.
-ENV MAX_WORKERS=10
+# Switch to the non-root user.
+USER appuser
 
-# Our Kubernetes does not allow serving on the default port 80.
-ENV PORT=8080
+# Expose the port.
 EXPOSE 8080
+
+# Run the application.
+CMD ["/start-api.sh"]
```

Comment:
@r-owen Have you tested a Docker image built with this dockerfile?

I'm trying to trace the Docker stages, and I think you might need to add some `COPY` steps to the `runtime-image` stage:

- start-api.sh
- alembic directory
- alembic.ini

Right now the only copy that's happening is the virtualenv from the install-image on line 56, but you'll also need to do a copy of these scripts because they won't be part of the Python virtualenv.

Reply:
Good catch. I compared it to the Dockerfile for timesquare and found that one difference: copying the start-api.sh script (plus my version of python was one dot version older). I updated it and the Dockerfile now looks exactly like timesquare.